File: app/src/main/java/com/vikingsen/cheesedemo/inject/AppModule.java
package com.vikingsen.cheesedemo.inject;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.preference.PreferenceManager;

import com.vikingsen.cheesedemo.model.database.AppDatabaseConfig;
import com.vikingsen.cheesedemo.util.SchedulerProvider;
import com.vikingsen.cheesedemo.util.SchedulerProvider.AppSchedulerProvider;

import org.dbtools.android.domain.config.DatabaseConfig;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
class AppModule {
    private final Application application;

    AppModule(Application application) {
        this.application = application;
    }

    @Provides
    @Singleton
    Application provideApplication() {
        return application;
    }

    @Provides
    @Singleton
    SharedPreferences provideSharedPreferences(Application application) {
        return PreferenceManager.getDefaultSharedPreferences(application);
    }

    @Provides
    @Singleton
    ConnectivityManager provideConnectivityManager(Application application) {
        return (ConnectivityManager) application.getSystemService(Context.CONNECTIVITY_SERVICE);
    }

    @Provides
    @Singleton
    DatabaseConfig provideDatabaseConfig(Application application) {
        return new AppDatabaseConfig(application);
    }

    @Provides
    @Singleton
    SchedulerProvider provideSchedulerProvider() {
        return new AppSchedulerProvider();
    }

}

File: app/src/main/java/com/vikingsen/cheesedemo/inject/AppModule.kt
package com.vikingsen.cheesedemo.inject

import android.app.Application
import android.content.Context
import android.content.SharedPreferences
import android.net.ConnectivityManager
import android.preference.PreferenceManager
import androidx.work.WorkManager
import com.vikingsen.cheesedemo.model.webservice.WebServiceModule
import com.vikingsen.cheesedemo.util.CoroutineContextProvider
import com.vikingsen.cheesedemo.util.CoroutineContextProvider.MainCoroutineContextProvider
import dagger.Module
import dagger.Provides
import dagger.Reusable
import javax.inject.Singleton

@Module(includes = [AssistedModule::class, WebServiceModule::class])
class AppModule(private val application: Application) {

    @Provides
    @Singleton
    fun provideApplication(): Application = application

    @Provides
    @Reusable
    fun provideSharedPreferences(application: Application): SharedPreferences = PreferenceManager.getDefaultSharedPreferences(application)

    @Provides
    @Reusable
    fun provideConnectivityManager(application: Application) = application.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    @Provides
    @Reusable
    fun provideCoroutineContextProvider(): CoroutineContextProvider = MainCoroutineContextProvider

    @Provides
    @Reusable
    fun provideWorkManager() = WorkManager.getInstance()

}
